src/cdk/a11y/aria-describer.ts

```diff
@@ -0,0 +1,135 @@
+import {Platform} from '@angular/cdk/platform';
+import {addAriaReferencedId, removeAriaReferencedId} from './aria-reference';
+
+/**
+ * Interface used to register message elements and keep a count of how many registrations have
+ * the same message and the reference to the message element used for the aria-describedby.
+ */
+export interface RegisteredMessage {
+  messageElement: HTMLElement;
+  hostElements: HTMLElement[];
+}
+
+/** ID used for the body container where all messages are appended. */
+export const MESSAGES_CONTAINER_ID = 'cdk-aria-describedby-messages';
+
+/** Global incremental identifier for each registered message element. */
+let nextId = 0;
+
+/** Global map of all registered message elements that have been placed into the document. */
+const registeredMessages = new Map<string, RegisteredMessage>();
+
+/** Container for all registered messages. */
+let messagesContainer: HTMLElement | null = null;
+
+/**
+ * Utility that creates visually hidden elements with a message content. Useful for elements that
+ * want to use aria-describedby to further describe themselves without adding additional visual
+ * content. Must be provided with as part of a component's injection tree so that it has access
+ * to the Renderer.
+ */
+@Injectable()
+export class AriaDescriber {
+  constructor(private _platform: Platform,
+              @Optional() private _renderer: Renderer2) {
+    if (!_renderer) {
+      throw Error('AriaDescriber must be provided through a component\'s provider decorator ' +
+          'so that it has access to the Angular Renderer');
+    }
+  }
+
+  /**
+   * Adds to the host element an aria-describedby reference to a hidden element that contains
+   * the message. If the same message has already been registered, then it will reuse the created
+   * message element.
+   */
+  addDescription(hostElement: HTMLElement, message: string) {
+    if (!this._platform.isBrowser || !message.trim()) { return; }
+
+    if (!registeredMessages.get(message)) {
+      const messageElement = this._createMessageElement(message);
+      registeredMessages.set(message, {messageElement, hostElements: []});
+    }
+
+    const registeredMessage = registeredMessages.get(message)!;
+    registeredMessage.hostElements.push(hostElement);
+    addAriaReferencedId(hostElement, registeredMessage.messageElement.id, 'aria-describedby');
+  }
+
+  /**
+   * Removes the host element's aria-describedby reference to the message element.
+   */
+  removeDescription(hostElement: HTMLElement, message: string) {
+    if (!this._platform.isBrowser || !message.trim()) { return; }
+
+    const registeredMessage = registeredMessages.get(message)!;
+    registeredMessage.hostElements = registeredMessage.hostElements.filter(el => el != hostElement);
+    removeAriaReferencedId(hostElement, registeredMessage.messageElement.id, 'aria-describedby');
+
+    if (registeredMessage.hostElements.length == 0) {
```

Comment:
Either triple equals or `!registeredMessage.hostElements.length` here and below.